S2198: Add C# (#1534)
SonarSource/rspec: rules/S2198/csharp/rule.adoc

Certain mathematical comparisons will always return the same value, and should simply not be made. 

> These comparisons will always return `false`:

Reviewer: ```suggestion
These comparisons will always return the same value (either always `true` or always `false` depending on the kind of comparison):
```

Certain mathematical comparisons will always return the same value, and should simply not be made. 

> These comparisons will always return ``++false++``:

Reviewer: ```suggestion
These comparisons will always return `false`:
```

> * comparing a `char` with a numeric constant that's outside the `char` range

Reviewer: The convention is to start with an upper letter and to end with a dot:
```suggestion
* Comparing a `char` with a numeric constant that's outside the `char` range.
```

> * comparing a `char` with a numeric constant that's outside the `char` range

Reviewer: Maybe better like so?
```suggestion
* Comparing a `char` with a numeric constant that is outside of the range of `char`.
```

> These comparisons will always return ``++false++``:
> 
> * comparing a ``++char++`` with a numeric constant that's outside the ``++char++`` range
> * comparing a ``++float++`` with a numeric constant that's outside the ``++float++`` range
> * comparing a ``++long++`` with a numeric constant that's outside the ``++long++`` range
> * comparing a ``++ulong++`` with a numeric constant that's outside the ``++ulong++`` range

Reviewer: ```suggestion
These comparisons will always return ``++false++``:

* comparing a `char` with a numeric constant that's outside the `char` range
* comparing a `float` with a numeric constant that's outside the `float` range
* comparing a `long` with a numeric constant that's outside the `long` range
* comparing a `ulong` with a numeric constant that's outside the `ulong` range
```

== Noncompliant Code Example

[source,text]
----

> float f = 42.0f;
> const double d = double.MaxValue;
> if (f <= d) { } // Noncompliant

Reviewer: Simpler:
```suggestion
float f = 42.0f;
if (f <= double.MaxValue) { } // Noncompliant
```
